Reset password flow and flash messages

kajalelohai opened the pull request:
Resolves #7

kajalelohai committed:
feat: send otp to reset password

kajalelohai committed:
fix: lint

kajalelohai committed:
Merge branch 'master' into feature/reset-password

bitspook commented on ui/src/duck/auth.ts:
```diff
@@ -28,6 +36,12 @@ export const fetchCurrentUser = createAsyncThunk<
+export const sendPasswordResetOTP = createAsyncThunk<
+  undefined,
+  SendOTPPayload,
+  { rejectValue: Error; state: PasswordResetOTP }
```
`createAsyncThunk` gets the state of the slice. So here it should be `AuthState` and not `PasswordResetOTP`.

bitspook commented on ui/src/entities/User.ts:
```diff
@@ -41,3 +45,14 @@ export const fetchCurrentUser = async (): Promise<User> => {
+  `;
+  const data = await client.request(query, payload);
+
+  return data ? data.sendPasswordResetOtp : undefined;
```
We don't need to check for absence of `data` this way. We can use [optional-chaining](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#optional-chaining) feature of typescript.

bitspook commented on ui/src/entities/User.ts:
```diff
@@ -41,3 +45,14 @@ export const fetchCurrentUser = async (): Promise<User> => {
   return data.me?.length ? data.me[0] : undefined;
 };
+
+export const sendPasswordResetOTP = async (payload: SendOTPPayload): Promise<undefined> => {
```
We don't need to create a new interface for accepting arguments of this function. It is (and will quite likely always be) a single string. We can save ourselves some boilerplate by making it `async (email: string)` instead. 

```suggestion
export const sendPasswordResetOTP = async (email: string): Promise<undefined> => {
```

kajalelohai committed:
feat: reset password

kajalelohai committed:
Merge branch 'master' into feature/reset-password

kajalelohai committed:
fix: remove export from .env

kajalelohai committed:
Merge remote-tracking branch 'origin/feature/reset-password' into feature/reset-password

kajalelohai committed:
fix: lint

kajalelohai committed:
refactor: remove ResetPassword component from LoginForm

kajalelohai committed:
refactor: get reset form show again

kajalelohai committed:
refactor: move dispatch to parent component

kajalelohai committed:
fix: lint

kajalelohai committed:
fix: Reset Password Modal css

Commit:
Merge branch 'master' into feature/reset-password

Commit:
fix: conflict resolution blunder

Commit:
fix: TextField taking care of where it is used

TextField shouldn't need to know where it is being used. Now, the way it works
is:
1. TextField has some default styles
2. TextField takes a `className` prop, which if provided, will take precedence
   over TextField's default styles. i.e TextField will not apply any styles if
   its user provides a class of her own

Fixes #33

Commit:
fix: styles for reset-password form

Commit:
feat: handle send-otp errors

Not showing the form error too elegantly yet. Gotta add those toast messages for that.

Commit:
feat: add FlashMessages!

Resolves #61

Commit:
refactor: remove uuid dependency

piggy-back redux-toolkit's requestId for async-thunks to provide ID for every
toast message

Commit:
feat: handle errors in send-otp

Commit:
feat: handle in-progress state of send-otp

Commit:
feat: create useFlash hook!

Commit:
feat: handle errors in reset-password

Commit:
feat: allow flashing strings

Commit:
fix: show proper and opaque error message for incorrect OTP

Commit:
feat: proper feedback on login success/fail

resolves #21

Commit:
feat: new styles for error flashes
